src/aiida/cmdline/params/types/path.py

```diff
@@ -78,24 +78,22 @@ def __init__(self, timeout_seconds=URL_TIMEOUT_SECONDS, **kwargs):
 
     def convert(self, value, param, ctx):
         """Overwrite `convert` Check first if `click.Path`-type, then check if URL."""
-        try:
-            return super().convert(value, param, ctx)
-        except click.exceptions.BadParameter:
-            return self.checks_url(value, param, ctx)
-
-    def checks_url(self, url, param, ctx):
-        """Check whether URL is reachable within timeout."""
         import socket
         import urllib.error
         import urllib.request
 
         try:
-            with urllib.request.urlopen(url, timeout=self.timeout_seconds):
+            with urllib.request.urlopen(value, timeout=self.timeout_seconds):
```

Comment:
Just wondering, are there any performance implications of calling urlopen first? How quickly does it fail for Paths?

Reply:
This was my original reason for not changing the order of checking, but I quickly checked and it seems really negligible.
```python
In [1]: import urllib.request

In [2]: def call_invalid():
   ...:     try:
   ...:         urllib.request.urlopen('some/path.txt')
   ...:     except ValueError:
   ...:         pass

In [3]: %timeit call_invalid
8.4 ns ± 0.231 ns per loop (mean ± std. dev. of 7 runs, 100,000,000 loops each)
```
As your investigation below showed, this is because it does a string match operation to check for a protocol which should be really fast and since filepaths will not match protocols, it will almost immediately error out and we move on to treating it as a local path.

Of course then the problem becomes the inverse. If the user _meant_ a URL but just forgot the protocol or had a typo, then they would get an error about the file not existing. I guess no matter how you turn it, it is going to be difficult to guess the intention. Maybe the best we can do is be explicit in when we tried both options and that we couldn't make it work either as a path nor a URL.

Reply:
Do we have an idea if one is likely than the other (path/URL), based where these options are used?

Reply:
This parameters is mostly used for the `--config` option. I think most users may have the config file locally, but support was added for URLs exactly because this makes sense for computer and code configurations which can be on servers, such as provided by the `aiida-code-registry`. I think I may revert the order to the original state and just improve the error message. If the local path validation fails, we still try the remote URL but if _that_ fails we make sure the error message is clear to state that it was tried as local path and then URL and both failed.

Reply:
That sounds like a good plan. I think you can still special-case the message if the path actually exists but we failed to read from it, as in your original fix before I started making a fuzz. 😅